File: typescript/tests/result-warnings.test.mjs
import assert from "node:assert/strict";
import test from "node:test";
import { createNodeYirClient } from "../dist/server/index.js";

test("HTTP parsing and polling retain optional result warnings without retrying generation", async () => {
  for (const availability of ["available", "expired"]) for (const warned of [false, true]) {
    const result = { availability, ...(availability === "available" ? { files: [{ url: "https://example.com/video.mp4", media_type: "video/mp4", expires_at: 1900000000 }] } : {}), ...(warned ? { warnings: ["additional_results_unavailable"] } : {}) };
    const payload = { object: "job", id: "1", model: "bytedance/seedance-2.0", status: "succeeded", result, error: null, created_at: 1 };
    let statusCalls = 0;
    let detailCalls = 0;
    const client = createNodeYirClient({ apiKey: "fixture", baseURL: "https://example.com", fetch: async (url, options) => {
      assert.equal(options.method, "GET");
      const pathname = new URL(url).pathname;
      if (pathname === "/v1/jobs/1/status") {
        statusCalls++;
        return new Response(JSON.stringify({ id: "1", status: "succeeded", error: null }), { headers: { "content-type": "application/json" } });
      }
      if (pathname === "/v1/jobs/1") {
        detailCalls++;
        return new Response(JSON.stringify(payload), { headers: { "content-type": "application/json" } });
      }
      throw new Error(`unexpected url: ${url}`);
    } });
    assert.deepEqual(await client.getJob("1"), payload);
    assert.deepEqual(await client.waitForJob("1"), payload);
    assert.equal(statusCalls, 1, "wait should poll status once");
    assert.equal(detailCalls, 2, "getJob once and terminal detail read once");
  }
});

File: go/result_warnings_test.go
package yir

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"reflect"
	"testing"
)

func TestResultWarningsHTTPAndWaitCompatibility(t *testing.T) {
	for _, availability := range []string{"available", "expired"} {
		for _, warned := range []bool{false, true} {
			result := map[string]any{"availability": availability}
			if availability == "available" {
				result["files"] = []any{map[string]any{"url": "https://example.com/video.mp4", "media_type": "video/mp4", "expires_at": 1900000000}}
			}
			if warned {
				result["warnings"] = []string{ResultWarningAdditionalResultsUnavailable}
			}
			statusCalls := 0
			detailCalls := 0
			server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
				if r.Method != "GET" {
					t.Error("unexpected mutation")
				}
				switch r.URL.Path {
				case "/v1/jobs/1/status":
					statusCalls++
					json.NewEncoder(w).Encode(map[string]any{"id": "1", "status": "succeeded", "error": nil})
				case "/v1/jobs/1":
					detailCalls++
					json.NewEncoder(w).Encode(map[string]any{"object": "job", "id": "1", "status": "succeeded", "result": result, "error": nil})
				default:
					t.Errorf("unexpected path: %s", r.URL.Path)
				}
			}))
			client, err := NewClient("fixture", ClientOptions{BaseURL: server.URL})
			if err != nil {
				t.Fatal(err)
			}
			for _, wait := range []bool{false, true} {
				var job Job
				if wait {
					job, err = client.WaitJob(context.Background(), "1", WaitOptions{})
				} else {
					job, err = client.GetJob(context.Background(), "1")
				}
				if err != nil || job.Status != "succeeded" || job.Result == nil {
					t.Fatalf("unexpected job: %+v %v", job, err)
				}
				var expected []string
				if warned {
					expected = []string{ResultWarningAdditionalResultsUnavailable}
				}
				if !reflect.DeepEqual(job.Result.Warnings, expected) {
					t.Fatal("warning lost")
				}
				raw, _ := json.Marshal(job.Result)
				var decoded map[string]any
				json.Unmarshal(raw, &decoded)
				if _, exists := decoded["warnings"]; exists != warned {
					t.Fatal("optional warning presence changed")
				}
				if availability == "available" && len(job.Result.Files) != 1 {
					t.Fatal("delivered files changed")
				}
			}
			server.Close()
			if statusCalls != 1 || detailCalls != 2 {
				t.Fatalf("unexpected calls: status=%d detail=%d", statusCalls, detailCalls)
			}
		}
	}
}
